src/Models/EventOptions.php

```diff
@@ -4,65 +4,62 @@
         if ($json) $this->set(json_decode($json, true));
     }
 
-    private function set($data)
-    {
-        if (isset($data[self::EVENT_USER])) {
-            $user = $data[self::EVENT_USER];
+    private function set($data){
+        $this->event = isset($data[self::EVENT]) ? $data[self::EVENT] : null;
+        $this->userId = isset($data[self::EVENT_USER_ID]) ? $data[self::EVENT_USER_ID] : null;
 
-            $this->user = new User(
-                isset($user[self::EVENT_USER_ID]) ? $user[self::EVENT_USER_ID] : '',
-                isset($user[self::EVENT_USER_NAME]) ? $user[self::EVENT_USER_NAME] : '',
-                isset($user[self::EVENT_USER_EMAIL]) ? $user[self::EVENT_USER_EMAIL] : ''
+        if (isset($data[self::EVENT_USER_TRAITS])) {
+            $userTraits = $data[self::EVENT_USER_TRAITS];
+            $this->userTraits = new UserTraits(
+                isset($userTraits[self::EVENT_USER_TRAITS_NAME]) ? $userTraits[self::EVENT_USER_TRAITS_NAME] : '',
+                isset($userTraits[self::EVENT_USER_TRAITS_EMAIL]) ? $userTraits[self::EVENT_USER_TRAITS_EMAIL] : '',
+                isset($userTraits[self::EVENT_USER_TRAITS_CREATED_AT]) ? $userTraits[self::EVENT_USER_TRAITS_CREATED_AT] : ''
             );
         }
-        $this->device = isset($data[self::EVENT_DEVICE]) ? $data[self::EVENT_DEVICE] : null;
-        $this->ip = isset($data[self::EVENT_IP]) ? $data[self::EVENT_IP] : null;
-        $this->userAgent = isset($data[self::EVENT_USER_AGENT]) ? $data[self::EVENT_USER_AGENT] : null;
-        $this->eventType = isset($data[self::EVENT_EVENT_TYPE]) ? $data[self::EVENT_EVENT_TYPE] : null;
-        $this->remoteIp = isset($data[self::EVENT_REMOTE_IP]) ? $data[self::EVENT_REMOTE_IP] : null;
 
-        if (isset($data[self::EVENT_PARAMS])) {
-            $this->setParams($data[self::EVENT_PARAMS]);
+        if (isset($data[self::EVENT_CONTEXT])) {
+            $context = $data[self::EVENT_CONTEXT];
+            $this->context = new Context(
+                isset($context[self::EVENT_CONTEXT_CLIENT_TOKEN]) ? $context[self::EVENT_CONTEXT_CLIENT_TOKEN] : '',
+                isset($context[self::EVENT_CONTEXT_IP]) ? $context[self::EVENT_CONTEXT_IP] : '',
+                isset($context[self::EVENT_CONTEXT_REMOTE_IP]) ? $context[self::EVENT_CONTEXT_REMOTE_IP] : '',
+                isset($context[self::EVENT_CONTEXT_HEADERS]) ? $context[self::EVENT_CONTEXT_HEADERS] : ''
+            );
         }
-    }
-
-    private function setParams($params)
-    {
-        $mappedParams = array();
 
-        if (isset($params)) {
-            for ($i = 1; $i <= self::EVENT_PARAMS_COUNT; $i++) {
-                // Map param 1 - COUNT if value exists in JSON
-                $mappedParams[self::EVENT_PARAMS_ATTR . $i] = isset($params[self::EVENT_PARAMS_ATTR . $i]) ? $params[self::EVENT_PARAMS_ATTR . $i] : '';
-            }
+        if (sizeof($data[self::EVENT_PROPERTIES] > self::MAX_EVENT_PROPERTIES_COUNT)) {
```

Comment:
sizeof counts array items, the parameter is invalid. 
Is this what you wanted?
`if ($data[self::EVENT_PROPERTIES] > self::MAX_EVENT_PROPERTIES_COUNT)`

Reply:
This is the correct line:
`if (sizeof($data[self::EVENT_PROPERTIES]) > self::MAX_EVENT_PROPERTIES_COUNT) {`